### PR TITLE
[clang] Extend diagnose_if to accept more detailed warning information

### PR DESCRIPTION
**Pull request by philnik777**
This implements parts of the extension proposed in https://discourse.llvm.org/t/exposing-the-diagnostic-engine-to-c/73092/7.

Specifically, this makes it possible to specify a diagnostic group in an optional third argument.

### CODE REVIEW

**Comment by philnik777** [comment]
This is missing a few tests and probably has some bugs, but I'd like to know whether this is a good approach before putting a lot of work into it.

**Comment by philnik777** [comment]
gentle ping~

File: clang-tools-extra/clang-tidy/ClangTidyDiagnosticConsumer.cpp [hunk 1]
```diff
@@ -177,17 +177,17 @@ DiagnosticBuilder ClangTidyContext::diag(
     StringRef CheckName, SourceLocation Loc, StringRef Description,
     DiagnosticIDs::Level Level /* = DiagnosticIDs::Warning*/) {
   assert(Loc.isValid());
-  unsigned ID = DiagEngine->getDiagnosticIDs()->getCustomDiagID(
-      Level, (Description + " [" + CheckName + "]").str());
+  unsigned ID = DiagEngine->getDiagnosticIDs()->getCustomDiagID(Level,
```
Comment (erichkeane) [inline clang-tools-extra/clang-tidy/ClangTidyDiagnosticConsumer.cpp:180]: Unrelated change?

File: clang-tools-extra/clang-tidy/ClangTidyDiagnosticConsumer.cpp [hunk 2]
```diff
@@ -177,17 +177,17 @@ DiagnosticBuilder ClangTidyContext::diag(
...
   CheckNamesByDiagnosticID.try_emplace(ID, CheckName);
   return DiagEngine->Report(Loc, ID);
 }
 
 DiagnosticBuilder ClangTidyContext::diag(
     StringRef CheckName, StringRef Description,
     DiagnosticIDs::Level Level /* = DiagnosticIDs::Warning*/) {
-  unsigned ID = DiagEngine->getDiagnosticIDs()->getCustomDiagID(
-      Level, (Description + " [" + CheckName + "]").str());
+  unsigned ID = DiagEngine->getDiagnosticIDs()->getCustomDiagID(Level,
```
Comment (erichkeane) [inline clang-tools-extra/clang-tidy/ClangTidyDiagnosticConsumer.cpp:189]: Same here.

File: clang/include/clang/Basic/DiagnosticIDs.h [hunk 3]
```diff
@@ -171,13 +172,61 @@ class DiagnosticMapping {
...
+              lhs.Class == rhs.Class &&
+              lhs.ShowInSystemHeader == rhs.ShowInSystemHeader &&
+              lhs.ShowInSystemMacro == rhs.ShowInSystemMacro &&
+              lhs.HasGroup == rhs.HasGroup &&
+              (!lhs.HasGroup || lhs.Group == rhs.Group) &&
+              lhs.Description == rhs.Description;
+    }
+
+    friend bool operator<(const CustomDiagDesc& lhs, const CustomDiagDesc& rhs) {
+      if (lhs.DefaultSeverity != rhs.DefaultSeverity)
```
Comment (erichkeane) [inline clang/include/clang/Basic/DiagnosticIDs.h:208]: Can you implement this with the 'std::tie' trick here?  As far as I can tell, thats what you're doing here.

File: clang/include/clang/Basic/DiagnosticIDs.h [hunk 4]
```diff
@@ -171,13 +172,61 @@ class DiagnosticMapping {
...
+    diag::Group Group = {};
+    std::string Description;
+
+    friend bool operator==(const CustomDiagDesc &lhs, const CustomDiagDesc &rhs) {
+      return lhs.DefaultSeverity == rhs.DefaultSeverity &&
+              lhs.Class == rhs.Class &&
+              lhs.ShowInSystemHeader == rhs.ShowInSystemHeader &&
+              lhs.ShowInSystemMacro == rhs.ShowInSystemMacro &&
+              lhs.HasGroup == rhs.HasGroup &&
+              (!lhs.HasGroup || lhs.Group == rhs.Group) &&
```
Comment (erichkeane) [inline clang/include/clang/Basic/DiagnosticIDs.h:203]: This little 'or' probably isn't necessary, since they'd both be default constructed, right?  So they'd be equal if both missing?

Also, std::tie trick would probably work here then.

That is:

`std::tie(lhs.DefaultSeverity, lhs.Class,...) == std::tie(rhs.DefaultSeverity, rhs.Class, ...)`

You can even create a private method of 'get-tied-version' so you could only type it 1x.

**Review by erichkeane — COMMENTED** [review]
I like this!  I can't really review the tidy stuff immediately, but I think this really makes sense.  You DO have some unrelated WS changes that should be removed though.

Also, Most of the changes this causes along the way I think are vast improvements.

File: clang/include/clang/Basic/DiagnosticIDs.h (hunk 4 above)
Comment (philnik777) [reply clang/include/clang/Basic/DiagnosticIDs.h]: I guess you're right. I've added a constructor to make sure it's actually an invariant. The `std::tie` didn't quite work, since it's binding to temporaries of the bitfields. I've made a value-tuple instead and that simplifies the code a lot. Thanks!

**Comment by erichkeane** [comment]
Please don't force push... it makes these 100x more difficult to review.  ALso, it seems that the latest push has broken quite a lot, according to the build bots.

File: clang/test/SemaCXX/diagnose_if-warning-group.cpp [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+
+#define diagnose_if(...) __attribute__((diagnose_if(__VA_ARGS__)))
+
+template <bool b>
+void diagnose_if_wcomma() diagnose_if(b, "oh no", "warning", "comma") {}
+
+template <bool b>
+void diagnose_if_wcomment() diagnose_if(b, "oh no", "warning", "comment") {}
+
+void bougus_warning() diagnose_if(true, "oh no", "warning", "bougus warning") {} // expected-error {{unknown warning group}}
```
Comment (AaronBallman) [inline clang/test/SemaCXX/diagnose_if-warning-group.cpp:11]: Perhaps `unknown warning group 'bogus warning'` instead? Mostly just because it may be hard to remember which one is the warning group, esp once we allow passing extra args for `%select` fields, etc.

File: clang/test/SemaCXX/diagnose_if-warning-group.cpp [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wcomment"
+  diagnose_if_wcomma<true>(); // expected-warning {{oh no}}
+  diagnose_if_wcomment<true>();
+#pragma clang diagnostic pop
+
+  diagnose_if_wcomma<true>(); // expected-warning {{oh no}}
+  diagnose_if_wcomment<true>(); // expected-error {{oh no}}
+
+  diagnose_if_wabi_default_error(); // expected-warning {{ABI stuff}}
```
Comment (AaronBallman) [inline clang/test/SemaCXX/diagnose_if-warning-group.cpp:40]: Do we want to allow `-Wno-error=foo` to work for diagnostics which are explicitly an error in the attribute? We don't let users do things like that for other errors and that suggests that `"error"` in `disable_if` is actually "warning that defaults to an error" rather than an error.

File: clang/test/SemaCXX/diagnose_if-warning-group.cpp [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+  diagnose_assume(); // expected-error {{Assume diagnostic}}
+
+  // Make sure that the -Wassume diagnostic isn't fatal
+  diagnose_if_wabi_default_error(); // expected-warning {{ABI stuff}}
+}
```
Comment (AaronBallman) [inline clang/test/SemaCXX/diagnose_if-warning-group.cpp:45]: Additional test coverage I'd like to see:

`diagnose_if(true, "oh no", "warning", "all")` -- Ensure `-Wall` enables the warning and `-Wno-all` disables it
`diagnose_if(true, "oh no", "warning", "pedantic")` -- Ensure `-Wpedantic` enables the warning, `-Wno-pedantic` disables it, and `-pedantic` does not turn it on (do we agree that `-Wpedantic` and `-pedantic` are different?).
`diagnose_if(true, "oh no", "warning", "format-nonliteral")` -- Ensure that `-Wformat` enables the warning and that `-Wno-format-nonliteral` and `-Wno-format` both disable the warning.

(And maybe coverage for pragma variants of the above cases?)

**Review by AaronBallman — COMMENTED** [review]
Sorry for the delay in review!

This seems like the right direction to go, in general. As @erichkeane, it's a bit hard to review due to the whitespace-only changes and changes to `__is_trivially_equality_comparable`, you should remove those changes or split them into a separate PR.

File: clang/test/SemaCXX/diagnose_if-warning-group.cpp (hunk 6 above)
Comment (philnik777) [reply clang/test/SemaCXX/diagnose_if-warning-group.cpp:40]: If you don't give the attribute a warning group, it should still produce an error. With a warning group it's currently implemented as a warning that defaults to an error, since it doesn't seem to make a ton of sense to add a warning group but then ask for it to never be a warning.

**Review by philnik777 — COMMENTED** [review]
> Sorry for the delay in review!

No worries! I dropped this for a few months myself. I can't expect to get a review within days when doing that :D

> This seems like the right direction to go, in general. As @erichkeane, it's a bit hard to review due to the whitespace-only changes and changes to `__is_trivially_equality_comparable`, you should remove those changes or split them into a separate PR.

Oops, looks like a commit got in here that was supposed to be part of another PR.

File: clang/test/SemaCXX/diagnose_if-warning-group.cpp (hunk 6 above)
Comment (AaronBallman) [reply clang/test/SemaCXX/diagnose_if-warning-group.cpp:40]: Ah, so the idea is:
```
diagnose_if(true, "oh no", "error" "thing") // #1
diagnose_if(true, "oh no", "error", "") // #2
diagnose_if(true, "oh no", "error") // #3
```
#1 is a warning that default to an error, #2 and #3 are errors with no ability to disable?

Comment (philnik777) [reply clang/test/SemaCXX/diagnose_if-warning-group.cpp:40]: I think #2 is currently rejected, but yeah, that's the idea.

Comment (AaronBallman) [reply clang/test/SemaCXX/diagnose_if-warning-group.cpp:40]: okay, that sounds reasonable to me (and I don't have strong opinions about supporting #2), thanks!

File: clang/test/SemaCXX/diagnose_if-warning-group.cpp (hunk 7 above)
Comment (AaronBallman) [reply clang/test/SemaCXX/diagnose_if-warning-group.cpp:63]: I tend to agree, but I believe there are intentional differences between `-pedantic` and `-Wpedantic` in terms of errors, see the discussion on: https://github.com/llvm/llvm-project/issues/84372

File: clang/include/clang/Basic/DiagnosticIDs.h [hunk 8]
```diff
@@ -101,91 +102,164 @@ namespace clang {
...
-    Result.WasUpgradedFromWarning = (Bits >> 3) & 1;
-    Result.Severity = Bits & 0x7;
-    return Result;
-  }
-
-  bool operator==(DiagnosticMapping Other) const {
-    return serialize() == Other.serialize();
-  }
-};
+  class DiagnosticMapping {
```
Comment (erichkeane) [inline clang/include/clang/Basic/DiagnosticIDs.h:105]: The reformatting of this file is troublesome for me... can you at least point out what the changes are here?  I'd prefer this not re-format the file (or if we want that, do it in a separate commit, then rebase this on top of that).

**Review by erichkeane — COMMENTED** [review]
I'm a little cold on review here, so I don't want to accept this, but I did a scroll through and don't have any problems here.  Additionally, a bunch of whitespace changes are still in this, which is making it difficult tor eview.

File: clang/include/clang/Basic/DiagnosticIDs.h (hunk 8 above)
Comment (philnik777) [reply clang/include/clang/Basic/DiagnosticIDs.h:105]: Except for `DiagnosticMapping` everything should be non-formatting changes. Not sure why I've reformatted this at all. Maybe I've reformatted something nearby any clang-format swallowed this too.

**Comment by philnik777** [comment]
@AaronBallman @erichkeane is there anything else to do before merging? It would be nice to merge this somewhat early in the release cycle, since I suspect this might be reverted a few times.

File: clang/include/clang/Basic/DiagnosticIDs.h [hunk 9]
```diff
@@ -179,13 +180,85 @@ class DiagnosticMapping {
...
+    diag::Severity DefaultSeverity : 3;
+    unsigned Class : 3;
+    unsigned ShowInSystemHeader : 1;
+    unsigned ShowInSystemMacro : 1;
+    unsigned HasGroup : 1;
+    diag::Group Group;
+    std::string Description;
+
+    auto get_as_tuple() const {
+      return std::tuple(DefaultSeverity, Class, ShowInSystemHeader,
```
Comment (erichkeane) [inline clang/include/clang/Basic/DiagnosticIDs.h:208]: nit: since this is just being used for comparison, would just a `std::tie` here be better?  That way the `string_view` isn't necessary?

**Review by erichkeane — APPROVED** [review]
I think this is fine to commit, sorry my review took so long!

File: clang/include/clang/Basic/DiagnosticIDs.h [hunk 10]
```diff
@@ -179,13 +180,85 @@ class DiagnosticMapping {
...
+          Description(std::move(Description)) {}
+
+    std::optional<diag::Group> GetGroup() const {
+      if (HasGroup)
+        return Group;
+      return std::nullopt;
+    }
+
+    diag::Severity GetDefaultSeverity() const { return DefaultSeverity; }
+    unsigned GetClass() const { return Class; }
```
Comment (AaronBallman) [inline clang/include/clang/Basic/DiagnosticIDs.h:232]: Should this return `Class` instead of `unsigned`?

File: clang/include/clang/Basic/DiagnosticIDs.h [hunk 11]
```diff
@@ -179,13 +180,85 @@ class DiagnosticMapping {
...
+
+    friend bool operator<(const CustomDiagDesc &lhs,
+                          const CustomDiagDesc &rhs) {
+      return lhs.get_as_tuple() < rhs.get_as_tuple();
+    }
+  };
+
+  struct GroupInfo {
+    diag::Severity Severity : 3;
+    bool HasNoWarningAsError : 1;
```
Comment (AaronBallman) [inline clang/include/clang/Basic/DiagnosticIDs.h:249]: `unsigned` and `LLVM_PREFERRED_TYPE` here as well.

File: clang/include/clang/Basic/DiagnosticIDs.h [hunk 12]
```diff
@@ -200,7 +273,33 @@ class DiagnosticIDs : public RefCountedBase<DiagnosticIDs> {
   // FIXME: Replace this function with a create-only facilty like
   // createCustomDiagIDFromFormatString() to enforce safe usage. At the time of
   // writing, nearly all callers of this function were invalid.
-  unsigned getCustomDiagID(Level L, StringRef FormatString);
+  unsigned getCustomDiagID(CustomDiagDesc Diag);
+
+  [[deprecated("Use a CustomDiagDesc instead of a Level")]] unsigned
```
Comment (AaronBallman) [inline clang/include/clang/Basic/DiagnosticIDs.h:278]: Similar concerns here about `-Werror` builds. Also, did clang-format cause this to be so horribly formatted?

File: clang/include/clang/Basic/DiagnosticIDs.h [hunk 13]
```diff
@@ -200,7 +273,33 @@ class DiagnosticIDs : public RefCountedBase<DiagnosticIDs> {
...
+  getCustomDiagID(Level Level, StringRef Message) {
+    return getCustomDiagID([&]() -> CustomDiagDesc {
+      switch (Level) {
+      case DiagnosticIDs::Level::Ignored:
+        return {diag::Severity::Ignored, std::string(Message), CLASS_WARNING,
```
Comment (AaronBallman) [inline clang/include/clang/Basic/DiagnosticIDs.h:283]: If it's ignored, why is it class warning and should be shown in system headers?

Actually, why are all of these saying to show in system headers? Was that the previous behavior and we're just matching it?

File: clang/lib/Basic/Diagnostic.cpp [hunk 14]
```diff
@@ -166,8 +166,12 @@ DiagnosticsEngine::DiagState::getOrAddMapping(diag::kind Diag) {
       DiagMap.insert(std::make_pair(Diag, DiagnosticMapping()));
 
   // Initialize the entry if we added it.
-  if (Result.second)
-    Result.first->second = DiagnosticIDs::getDefaultMapping(Diag);
+  if (Result.second) {
+    Result.first->second = DiagIDs.getDefaultMapping(Diag);
+    if (DiagnosticIDs::IsCustomDiag(Diag)) {
+      DiagIDs.initCustomDiagMapping(Result.first->second, Diag);
+    }
```
Comment (AaronBallman) [inline clang/lib/Basic/Diagnostic.cpp:173]:
```suggestion
    if (DiagnosticIDs::IsCustomDiag(Diag))
      DiagIDs.initCustomDiagMapping(Result.first->second, Diag);
```

File: clang/lib/Basic/DiagnosticIDs.cpp [hunk 15]
```diff
@@ -269,11 +268,60 @@ CATEGORY(INSTALLAPI, REFACTORING)
...
+// Custom Diagnostic information
+//===----------------------------------------------------------------------===//
+
+namespace clang {
+namespace diag {
+using CustomDiagDesc = DiagnosticIDs::CustomDiagDesc;
+class CustomDiagInfo {
+  std::vector<CustomDiagDesc> DiagInfo;
+  std::map<CustomDiagDesc, unsigned> DiagIDs;
+  std::map<diag::Group, std::vector<unsigned>> GroupToDiags;
```
Comment (AaronBallman) [inline clang/lib/Basic/DiagnosticIDs.cpp:281]: Should these be using LLVM ADTs instead for better performance? Specifically, `SmallVector`, `[Small]DenseMap`, and `MapVector`?

File: clang/lib/Basic/DiagnosticIDs.cpp [hunk 16]
```diff
@@ -286,6 +334,18 @@ DiagnosticMapping DiagnosticIDs::getDefaultMapping(unsigned DiagID) {
   return Info;
 }
 
+void DiagnosticIDs::initCustomDiagMapping(DiagnosticMapping &Mapping,
+                                          unsigned DiagID) {
+  assert(IsCustomDiag(DiagID));
+  const auto &Diag = CustomDiagInfo->getDescription(DiagID);
+  if (auto Group = Diag.GetGroup()) {
+    auto GroupInfo = GroupInfos[static_cast<size_t>(*Group)];
```
Comment (AaronBallman) [inline clang/lib/Basic/DiagnosticIDs.cpp:342]: Not a good use of `auto` because the type is not particularly obvious from the initialization.

File: clang/lib/Basic/DiagnosticIDs.cpp [hunk 17]
```diff
@@ -509,30 +499,32 @@ DiagnosticIDs::getDiagnosticLevel(unsigned DiagID, SourceLocation Loc,
 diag::Severity
 DiagnosticIDs::getDiagnosticSeverity(unsigned DiagID, SourceLocation Loc,
                                      const DiagnosticsEngine &Diag) const {
-  assert(getBuiltinDiagClass(DiagID) != CLASS_NOTE);
+  const bool IsCustomDiag = DiagnosticIDs::IsCustomDiag(DiagID);
```
Comment (AaronBallman) [inline clang/lib/Basic/DiagnosticIDs.cpp:502]: We don't typically use top-level const on locals (here and elsewhere in the patch).

File: clang/lib/Basic/DiagnosticIDs.cpp [hunk 18]
```diff
@@ -597,6 +592,15 @@ DiagnosticIDs::getDiagnosticSeverity(unsigned DiagID, SourceLocation Loc,
   return Result;
 }
 
+DiagnosticIDs::Class DiagnosticIDs::getDiagClass(unsigned DiagID) const {
+  if (IsCustomDiag(DiagID))
+    return Class(CustomDiagInfo->getDescription(DiagID).GetClass());
+
+  if (const StaticDiagInfoRec *Info = GetDiagInfo(DiagID))
+    return Class(Info->Class);
+  return Class(~0U);
```
Comment (AaronBallman) [inline clang/lib/Basic/DiagnosticIDs.cpp:601]: This is UB because it's outside the range of representable values for the `Class` enum; maybe make it an enum member instead?

File: clang/lib/StaticAnalyzer/Core/TextDiagnostics.cpp [hunk 19]
```diff
@@ -91,7 +91,7 @@ class TextDiagnostics : public PathDiagnosticConsumer {
                                     ? " [" + PD->getCheckerName() + "]"
                                     : "")
                                    .str();
-
+      // assert(!DiagEng.isIgnored(WarnID, PD->getLocation().asLocation()));
```
Comment (AaronBallman) [inline clang/lib/StaticAnalyzer/Core/TextDiagnostics.cpp:94]: Can be removed?

File: clang/tools/libclang/CXStoredDiagnostic.cpp [hunk 20]
```diff
@@ -33,14 +33,14 @@ CXDiagnosticSeverity CXStoredDiagnostic::getSeverity() const {
     case DiagnosticsEngine::Error:   return CXDiagnostic_Error;
     case DiagnosticsEngine::Fatal:   return CXDiagnostic_Fatal;
   }
-  
+
```
Comment (AaronBallman) [inline clang/tools/libclang/CXStoredDiagnostic.cpp:36]: A lot of unrelated formatting changes (feel free to land as NFC and rebase on top of it if you want, but this keeps blames cleaner).

**Review by AaronBallman — COMMENTED** [review]
Sorry for the delayed review! Just some minor points, mostly.

File: clang/include/clang/Basic/Diagnostic.h [hunk 21]
```diff
@@ -870,7 +872,8 @@ class DiagnosticsEngine : public RefCountedBase<DiagnosticsEngine> {
   /// \param FormatString A fixed diagnostic format string that will be hashed
   /// and mapped to a unique DiagID.
   template <unsigned N>
-  unsigned getCustomDiagID(Level L, const char (&FormatString)[N]) {
+  [[deprecated("Use a CustomDiagDesc instead of a Level")]] unsigned
```
[the remark this answers is not in the document]
Comment (philnik777) [reply clang/include/clang/Basic/Diagnostic.h:875]: I guess so. Should we maybe enable `-Werror` in the pre-commit CI to catch these kinds of problems?

File: clang/include/clang/Basic/DiagnosticIDs.h (hunk 13 above)
Comment (philnik777) [reply clang/include/clang/Basic/DiagnosticIDs.h:283]: Yes, this is just matching the previous behaviour. I agree that it doesn't make a ton of sense in some cases, but this should go at some point anyways, so I didn't want to mess with it.

File: clang/include/clang/Basic/Diagnostic.h (hunk 21 above)
[the remark this answers is not in the document]
Comment (AaronBallman) [reply clang/include/clang/Basic/Diagnostic.h]: If someone wanted to explore enabling that, it might not be a bad idea tbh.

**Review by AaronBallman — APPROVED** [review]
LGTM!

**Comment by kazutakahirata** [comment]
The build bots seem to be complaining about the same spot:

```
flang/lib/Frontend/TextDiagnosticPrinter.cpp:42:29: error: call to non-static member function without an object argument
      clang::DiagnosticIDs::getWarningOptionForDiag(info.getID());
```

Maybe you need to update the call to `getWarningOptionForDiag`?

**Comment by mikaelholmen** [comment]
Hello,

I noticed that before this patch
```clang empty.c -fsanitize=undefined -ffine-grained-bitfield-accesses -c -Werror```
(on empty file empty.c) resulted in
```warning: option '-ffine-grained-bitfield-accesses' cannot be enabled together with a sanitizer; flag ignored```
but with the patch it results in
```error: option '-ffine-grained-bitfield-accesses' cannot be enabled together with a sanitizer; flag ignored [-Werror]```

Is this on purpose?

(I originally asked this in
 https://github.com/llvm/llvm-project/pull/108453#issuecomment-2352155766
but I suppose this PR has a bit wider audience.)

**Comment by AaronBallman** [comment]
I think there are multiple bugs here, at least one seems to be preexisting. It seems that perhaps diagnostics issued early enough in the driver do not seem to honor diagnostic groups: https://godbolt.org/z/sh4e9noe1

Note how it's missing `[-Woption-ignored]` like you get in: https://godbolt.org/z/xvn59hEva

However, note how we're missing that diagnostic *entirely* now: https://godbolt.org/z/znr63dj6o

**Comment by whentojump** [comment]
> Hello,
> 
> I noticed that before this patch `clang empty.c -fsanitize=undefined -ffine-grained-bitfield-accesses -c -Werror` (on empty file empty.c) resulted in `warning: option '-ffine-grained-bitfield-accesses' cannot be enabled together with a sanitizer; flag ignored` but with the patch it results in `error: option '-ffine-grained-bitfield-accesses' cannot be enabled together with a sanitizer; flag ignored [-Werror]`
> 
> Is this on purpose?
> 
> (I originally asked this in [#108453 (comment)](https://github.com/llvm/llvm-project/pull/108453#issuecomment-2352155766) but I suppose this PR has a bit wider audience.)

We are seeing similar changes in MC/DC-related warnings. Steps to reproduce:

```shell
cat > test.c << 'EOF'
int foo(int x) { return x; }
int main(void) {
    int a, b, c, d, e, f, g, h;
    a && b && c && d && e && f && g;
    a && foo( b && c );
    return 0;
}
EOF
clang -Xclang -fmcdc-max-conditions=6 -Werror -Wno-unused-value -fprofile-instr-generate -fcoverage-mapping -fcoverage-mcdc test.c -o test
```

File: clang/lib/Basic/DiagnosticIDs.cpp [hunk 22]
```diff
@@ -489,13 +485,7 @@ static DiagnosticIDs::Level toLevel(diag::Severity SV) {
 DiagnosticIDs::Level
 DiagnosticIDs::getDiagnosticLevel(unsigned DiagID, SourceLocation Loc,
                                   const DiagnosticsEngine &Diag) const {
-  // Handle custom diagnostics, which cannot be mapped.
```
Comment (kadircet) [inline clang/lib/Basic/DiagnosticIDs.cpp:492]: so despite this change making sense at a high level, this is a big behavior change that isn't really obvious in context of this PR.

before this change, clang didn't propagate `-Werror` like mappings into custom diagnostics. Hence some compilations that succeed now with custom warning (clang-plugins etc.), will start failing after this change as those will now turn into errors.
We're already seeing this internally and both https://github.com/llvm/llvm-project/pull/70976#issuecomment-2357811301 and https://github.com/llvm/llvm-project/pull/70976#issuecomment-2372463357 are mentioning this as well.

Since this is triggering a backward incompatible behavior, can we have this bit reverted and decide how/whether we're planning to move forward with this?

Comment (bjope) [reply clang/lib/Basic/DiagnosticIDs.cpp:492]: > Since this is triggering a backward incompatible behavior, can we have this bit reverted and decide how/whether we're planning to move forward with this?

Afaict the author ( @philnik777 ) has neither responded to the post-commit comments/question here, neither in the PR related to the reapply at https://github.com/llvm/llvm-project/pull/108453#issuecomment-2352155766 , and I don't know much about this to understand if there is an easy fix. But if we are suspecting "multiple bugs here" then maybe someone should just go ahead and revert this again.

Btw, it seems like the reapply already has been reverted once (https://github.com/llvm/llvm-project/commit/e1bd9740faa62c11cc785a7b70ec1ad17e286bd1) and then reapplied a second time (https://github.com/llvm/llvm-project/commit/e39205654dc11c50bd117e8ccac243a641ebd71f). And there are comments here https://github.com/llvm/llvm-project/commit/e39205654dc11c50bd117e8ccac243a641ebd71f#commitcomment-146720911 about further fixups that might cause conflicts if reverting again.

PS. The commit message in https://github.com/llvm/llvm-project/commit/e39205654dc11c50bd117e8ccac243a641ebd71f is not very informative as it says that the commit msg body says that it is a revert, while it actually is a re-apply. Maybe that is another reason to revert this mess and reapply it using a new PR with an appropriate commit message describing the feature.

Comment (fmayer) [reply clang/lib/Basic/DiagnosticIDs.cpp:492]: I agree, let's revert if this is causing problems.

Comment (kadircet) [reply clang/lib/Basic/DiagnosticIDs.cpp:492]: ok, putting together a revert for https://github.com/llvm/llvm-project/commit/e39205654dc11c50bd117e8ccac243a641ebd71f

**Comment by mikaelholmen** [comment]
Thanks @kadircet
